From 8954e9a6da329f2265f22fee305f60d4c511dea0 Mon Sep 17 00:00:00 2001
From: Lim Siew Hoon <siew.hoon.lim@intel.com>
Date: Mon, 31 Jul 2023 16:50:37 +0800
Subject: [PATCH 34/77] video-info-dma: add
 gst_video_info_dma_drm_from_video_info()

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4680>

Backport from upstream:
https://gitlab.freedesktop.org/gstreamer/gstreamer/-/commit/1477ed1ae5bc206d9f5e7dfe3a54b502ad0fca91

Signed-off-by: Lim Siew Hoon <siew.hoon.lim@intel.com>
---
 gst-libs/gst/video/video-info-dma.c | 44 +++++++++++++++++++++++++++++
 gst-libs/gst/video/video-info-dma.h |  6 ++++
 tests/check/libs/video.c            |  6 ++++
 3 files changed, 56 insertions(+)

diff --git a/gst-libs/gst/video/video-info-dma.c b/gst-libs/gst/video/video-info-dma.c
index ec0e500e..2ec08511 100644
--- a/gst-libs/gst/video/video-info-dma.c
+++ b/gst-libs/gst/video/video-info-dma.c
@@ -400,6 +400,50 @@ gst_video_info_dma_drm_new_from_caps (const GstCaps * caps)
   }
 }
 
+/**
+ * gst_video_info_dma_drm_from_video_info:
+ * @drm_info: (out caller-allocates): #GstVideoInfoDmaDrm
+ * @info: a #GstVideoInfo
+ * @modifier: the associated modifier value.
+ *
+ * Fills @drm_info if @info's format has a valid drm format and @modifier is also
+ * valid
+ *
+ * Returns: %TRUE if @drm_info is filled correctly.
+ *
+ * Since: 1.24
+ */
+gboolean
+gst_video_info_dma_drm_from_video_info (GstVideoInfoDmaDrm * drm_info,
+    const GstVideoInfo * info, guint64 modifier)
+{
+  GstVideoFormat format;
+  guint32 fourcc;
+
+  g_return_val_if_fail (drm_info != NULL, FALSE);
+  g_return_val_if_fail (info != NULL, FALSE);
+
+  if (modifier == DRM_FORMAT_MOD_INVALID)
+    return FALSE;
+  format = GST_VIDEO_INFO_FORMAT (info);
+  fourcc = gst_video_dma_drm_fourcc_from_format (format);
+  if (fourcc == DRM_FORMAT_INVALID)
+    return FALSE;
+
+  drm_info->vinfo = *info;
+  drm_info->drm_fourcc = fourcc;
+  drm_info->drm_modifier = modifier;
+
+  /* no need to change format to GST_VIDEO_INFO_ENCODED since its modifier is
+   * linear */
+  if (modifier == DRM_FORMAT_MOD_LINEAR)
+    return TRUE;
+
+  return gst_video_info_set_interlaced_format (&drm_info->vinfo,
+      GST_VIDEO_FORMAT_ENCODED, GST_VIDEO_INFO_INTERLACE_MODE (info),
+      GST_VIDEO_INFO_WIDTH (info), GST_VIDEO_INFO_HEIGHT (info));
+}
+
 /**
  * gst_video_dma_drm_fourcc_from_string:
  * @format_str: a drm format string
diff --git a/gst-libs/gst/video/video-info-dma.h b/gst-libs/gst/video/video-info-dma.h
index ef6c198a..61f8319e 100644
--- a/gst-libs/gst/video/video-info-dma.h
+++ b/gst-libs/gst/video/video-info-dma.h
@@ -69,6 +69,12 @@ GstCaps *            gst_video_info_dma_drm_to_caps       (const GstVideoInfoDma
 GST_VIDEO_API
 gboolean             gst_video_info_dma_drm_from_caps     (GstVideoInfoDmaDrm * drm_info,
                                                            const GstCaps * caps);
+
+GST_VIDEO_API
+gboolean             gst_video_info_dma_drm_from_video_info
+                                                          (GstVideoInfoDmaDrm * drm_info,
+                                                           const GstVideoInfo * info,
+                                                           guint64 modifier);
 GST_VIDEO_API
 GstVideoInfoDmaDrm * gst_video_info_dma_drm_new_from_caps (const GstCaps * caps);
 
diff --git a/tests/check/libs/video.c b/tests/check/libs/video.c
index b2becebe..1c6049ed 100644
--- a/tests/check/libs/video.c
+++ b/tests/check/libs/video.c
@@ -4160,6 +4160,12 @@ GST_START_TEST (test_info_dma_drm)
   gst_caps_unref (caps);
   gst_caps_unref (ncaps);
 
+  fail_unless (gst_video_info_dma_drm_from_video_info (&info, &vinfo, 0));
+  fail_unless (GST_VIDEO_INFO_FORMAT (&info.vinfo) == GST_VIDEO_FORMAT_NV12);
+
+  fail_unless (gst_video_info_dma_drm_from_video_info (&info, &vinfo,
+          0x100000000000002));
+  fail_unless (GST_VIDEO_INFO_FORMAT (&info.vinfo) == GST_VIDEO_FORMAT_ENCODED);
 }
 
 GST_END_TEST;
-- 
2.40.1

